wallet/src/Pos/Util/Mnemonic.hs

```diff
@@ -0,0 +1,664 @@
+
+-- | Entropy as a non-empty sequence of bytes, multiple of 4 bytes
+newtype Entropy = Entropy ByteString deriving (Eq, Show, Generic)
+
+-- Type Alias for readability. A bit can only be 1 or 0, but we
+-- use Word8 instead of Bool to leverage some available type-classes.
+type Bit = Word8
+
+-- Type Alias for readability
+type Word11 = Int
+
+
+-- | Initial seed has to be vector or length multiple of 4 bytes and shorter
+-- than 64 bytes. Not that this is good for testing or examples, but probably
+-- not for generating truly random Mnemonic words.
+--
+-- See 'Crypto.Random.Entropy (getEntropy)'
+instance Arbitrary Entropy where
+    arbitrary =
+        Entropy . B8.pack <$> oneof [ vectorOf (4 * n) arbitrary | n <- [1..16] ]
+
+-- FIXME: Suggestion, we could -- when certain flags are turned on -- display
+-- a fingerprint of the Mnemonic, like a PKBDF2 over n iterations. This could be
+-- useful for debug to know whether two users are using the same mnemonic words
+-- and relatively benign EVEN THOUGH, it will permit to somewhat tight requests
+-- to a specific identity (since mnemonic words are 'unique', they are supposed
+-- to uniquely identify users, hence the privacy issue). For debbugging only and
+-- with the user consent, that's something we could do.
+instance Buildable Mnemonic where
+    build _ =
+        "<mnemonic>"
+
+instance Buildable (SecureLog Mnemonic) where
+    build _ =
+        "<mnemonic>"
+
+-- | To use everytime we need to show an example of a Mnemonic. This particular
+-- mnemonic is rejected to prevent users from using it on a real wallet.
+instance Default Mnemonic where
+    def = Mnemonic
+        { getEntropy = Entropy "\211\177\US\"\245\163\233\152\169]\149\DC3\239)\234\172"
+        , getWords =
+            [ (1693,"squirrel")
+            , (1095,"material")
+            , (1605,"silly")
+            , (1882,"twice")
+            , (500,"direct")
+            , (1634,"slush")
+            , (1323,"pistol")
+            , (1429,"razor")
+            , (159,"become")
+            , (970,"junk")
+            , (981,"kingdom")
+            , (710,"flee")
+            ]
+        }
+
+-- Same remark from 'Arbitrary Entropy' applies here.
+instance Arbitrary Mnemonic where
+    arbitrary =
+        entropyToMnemonic <$> arbitrary
+
+instance FromJSON Mnemonic where
+    parseJSON =
+        parseJSON >=> (eitherToParser . mkMnemonic)
+      where
+        eitherToParser :: Either Text Mnemonic -> Parser Mnemonic
+        eitherToParser =
+            either (fail . show) pure
+
+instance ToJSON Mnemonic where
+    toJSON =
+        toJSON . map snd . getWords
+
+instance ToSchema Mnemonic where
+    declareNamedSchema _ = do
+        NamedSchema _ schema <- declareNamedSchema (Proxy @[Text])
+        return $ NamedSchema (Just "Mnemonic") schema
+
+
+-- | Smart-constructor for the Entropy
+mkEntropy :: ByteString -> Either Text Entropy
+mkEntropy bytes = do
+    when (BS.length bytes <= 0) $
```

Comment:
I totally agree with you and... yes.. we diverge from the official spec quite a lot (especially when it comes to generating the seed). 
For now, I kept the behavior the same to avoid any issue and not introduce regression although I believe the range could be reduced indeed.